### PR TITLE
[libsai] Implement APIs set/get attribute

### PR DESCRIPTION
**Pull request by jimmyzhai**
Following https://github.com/sonic-net/DASH/pull/648, continue to refactor libsai implementation for easier read and maintenance:

1. Avoid using compilated j2 template as best
2. Use c++ code directly to express the logic of API implementation, based on p4 table metadata info (p4meta.[h, cpp])
3. Implement APIs set/get attribute by using common `DashSai::set` and `DashSai::get`, which fix the issue https://github.com/sonic-net/DASH/issues/650
4. Update libsai test (vnet_out.cpp) to cover sanity test of APIs set/get attribute

Note: To refactor APIs create/remove with current approach in next PR. 

The generated sample code of APIs set/get attribute is as below:

- SAI object objectID
image
p4meta `outbound_routing_group_meta_table` definition
image

- SAI object entry
image
p4meta `outbound_routing_entry_meta_table`  definition
image

### CODE REVIEW

**Comment by r12f** [comment]
Hi Junhua, do you mind to update the PR description on how the generated code looks like? It will help people to understand this PR better.

File: dash-pipeline/SAI/src/p4meta.h [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+            _In_ const V &value,
+            _Inout_ p4::v1::FieldMatch_LPM *t)
+    {
+        assert (field == "ipPrefix");
+        return ipPrefixSetVal(value, t, bitwidth);
+    }
+
+    template<typename T>
+    static inline
+    void get_value(
```
Comment (r12f) [inline dash-pipeline/SAI/src/p4meta.h:148]: better rename the functions. the names are too vague.

File: dash-pipeline/SAI/src/p4meta.h [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+    }
+
+    static inline
+    void get_value(
+            _In_ const std::string &field,
+            _In_ uint32_t bitwidth,
+            _In_ const p4::v1::FieldMatch_LPM *t,
+            _Out_ sai_attribute_value_t &value)
+    {
+        assert (field == "ipPrefix");
```
Comment (r12f) [inline dash-pipeline/SAI/src/p4meta.h:206]: move non-template to p4meta.cpp

**Comment by jimmyzhai** [comment]
> Hi Junhua, do you mind to update the PR description on how the generated code looks like? It will help people to understand this PR better.

Added generated sample code in PR description

File: dash-pipeline/SAI/src/p4meta.h (hunk 1 above)
Comment (jimmyzhai) [reply dash-pipeline/SAI/src/p4meta.h]: fixed. Update functions to have a meaningful name.

File: dash-pipeline/SAI/src/p4meta.h (hunk 2 above)
Comment (jimmyzhai) [reply dash-pipeline/SAI/src/p4meta.h]: fixed.

**Comment by r12f** [comment]
> > Hi Junhua, do you mind to update the PR description on how the generated code looks like? It will help people to understand this PR better.
> 
> Added generated sample code in PR description

Thanks for sharing the details out! This makes the review much easier!

File: dash-pipeline/tests/libsai/vnet_out/vnet_out.cpp [hunk 3]
```diff
@@ -231,6 +241,12 @@ int main(int argc, char **argv)
         return 1;
     }
 
+    attr.id = SAI_ENI_ATTR_VM_UNDERLAY_DIP;
+    attr.value.ipaddr.addr = {0};
+    status = dash_eni_api->get_eni_attribute(eni_id, 1, &attr);
```
Comment (r12f) [inline dash-pipeline/tests/libsai/vnet_out/vnet_out.cpp:246]: thanks for adding the tests!

File: dash-pipeline/SAI/src/dashsai.cpp [hunk 4]
```diff
@@ -705,6 +752,40 @@ bool DashSai::removeFromTable(
...
+    if (!m_apiInitialized)
+    {
+        DASH_LOG_ERROR("api not initialized");
+
+        return false;
+    }
+
+    MUTEX;
+
+    auto range = m_tableEntryMap.equal_range(id);
```
Comment (r12f) [inline dash-pipeline/SAI/src/dashsai.cpp:770]: emmmm, do we really need a multi-map here? each object id should be unique with the object id encoding, am I right? @kcudnik

File: dash-pipeline/SAI/src/p4meta.cpp [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+            }
+            else if (mf->field_id() == meta_key->ip_is_v6_field_id) {
+                pair_key.second = mf;
+            }
+        }
+
+        return pair_key;
+    }
+
+    std::pair<p4::v1::Action_Param*, p4::v1::Action_Param*> get_action_param_pair_from_p4_table_entry(
```
Comment (r12f) [inline dash-pipeline/SAI/src/p4meta.cpp:333]: get_action_param_with_is_v6_flag_from_p4_table_entry

Comment (r12f) [reply dash-pipeline/SAI/src/p4meta.cpp:333]: also use out parameter will be more explicit, the pair.second is very unintuitive to read.

File: dash-pipeline/SAI/src/p4meta.cpp [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+            }
+            else if (param->param_id() == meta_param->ip_is_v6_field_id) {
+                pair_param.second = param;
+            }
+        }
+
+        return pair_param;
+    }
+
+    bool string_has_suffix(const std::string &str, const std::string &suffix)
```
Comment (r12f) [inline dash-pipeline/SAI/src/p4meta.cpp:353]: string_ends_with

**Review by r12f — APPROVED** [review]
looks good to me! commends can be solved in the next PR, all minor and non-blocking.

File: dash-pipeline/SAI/src/dashsai.cpp [hunk 7]
```diff
@@ -705,6 +752,40 @@ bool DashSai::removeFromTable(
     return retCode == grpc::StatusCode::OK;
 }
 
+bool DashSai::getFromTable(
+        _In_ sai_object_id_t id,
+        _Out_ std::shared_ptr<p4::v1::TableEntry> &entry)
+{
+    DASH_LOG_ENTER();
+
+    if (!m_apiInitialized)
```
[the remark this answers is not in the document]
Comment (jimmyzhai) [reply dash-pipeline/SAI/src/dashsai.cpp:761]: fix it at https://github.com/sonic-net/DASH/pull/653

File: dash-pipeline/SAI/src/dashsai.cpp (hunk 4 above)
Comment (jimmyzhai) [reply dash-pipeline/SAI/src/dashsai.cpp:770]: @r12f , `map` is enough. fix it at https://github.com/sonic-net/DASH/pull/653

File: dash-pipeline/SAI/src/p4meta.cpp (hunk 5 above)
Comment (jimmyzhai) [reply dash-pipeline/SAI/src/p4meta.cpp:333]: fix it at https://github.com/sonic-net/DASH/pull/653

File: dash-pipeline/SAI/src/p4meta.cpp (hunk 6 above)
Comment (jimmyzhai) [reply dash-pipeline/SAI/src/p4meta.cpp:353]: fix it at https://github.com/sonic-net/DASH/pull/653

**Pull request merged**